rplugin/python3/LanguageClient/LanguageClient.py

```diff
@@ -983,14 +1013,33 @@ def addHighlightAndSign(self, params):
         cmd = getCommandUpdateSigns(self.signs, signs)
         self.signs = signs
         self.asyncCommand(cmd)
+        old_list = {frozenset(item.items()) for item in self.get_errorlist()}
+        new_list = {frozenset(item.items()) for item in qflist}
+        if not old_list or old_list != new_list:
+            self.set_errorlist(qflist)
+            if self.nvim.funcs.mode() == 'i':
+                self.open = True
+            else:
+                self.open_or_close_errors(from_autocmd=False)
 
-        if self.diagnosticsList == "quickfix":
-            self.nvim.funcs.setqflist(qflist, "r")
-        elif self.diagnosticsList == "location":
-            self.nvim.funcs.setloclist(0, qflist, "r")
+    @neovim.autocmd('InsertLeave', pattern='*')
+    def open_or_close_errors(self, from_autocmd=True):
+        if from_autocmd and not self.open:
+            return
+        self.open = False
+        prev_winnr = self.nvim.funcs.winnr()
+        messages_nr = len(self.get_errorlist())
+        if messages_nr:
+            self.open_errorlist(messages_nr)
+        else:
+            self.close_errorlist()
+        if prev_winnr != self.nvim.funcs.winnr():
+            self.nvim.command(str(prev_winnr) + ' wincmd w')
 
     @neovim.autocmd("CursorMoved", pattern="*", eval="line('.')")
     def handleCursorMoved(self, line) -> None:
+        if self.nvim.current.buffer.options['buftype']:
```

Comment:
👍  But please make a comment. It's not easy to understand what it does in first glance.